src/ensembl/src/content/app/custom-download/components/content-builder/ContentBuilder.tsx

```diff
@@ -0,0 +1,263 @@
+import JSONValue, { PrimitiveOrArrayValue } from 'src/shared/types/JSON';
+
+type Path = (string | number)[];
+let path: Path = [];
+
+type ContentBuilderProps = {
+  data: AttributeWithContent;
+  selectedData: JSONValue;
+  onChange: (type: string, path: Path, payload: PrimitiveOrArrayValue) => void;
+  contentState: JSONValue;
+  onContentStateChange: (
+    type: string,
+    path: Path,
+    payload: PrimitiveOrArrayValue
+  ) => void;
+  path?: Path;
+  contentProps?: {
+    [key: string]: JSONValue;
+  };
+};
+
+const ContentBuilder = (props: ContentBuilderProps) => {
+  const onChangeHandler = (
+    type: string,
+    path: Path,
+    payload: PrimitiveOrArrayValue
+  ) => {
+    props.onChange(type, path, payload);
+  };
+
+  const onContentStateChangeHandler = (
+    type: string,
+    path: Path,
+    payload: PrimitiveOrArrayValue
+  ) => {
+    props.onContentStateChange(type, path, payload);
+  };
+
+  const buildCheckboxWithSelect = (entry: AttributeWithOptions, path: Path) => {
+    const currentPath = [...path, entry.id];
+
+    const selectedOptions = get(props.selectedData, currentPath, []);
+
+    return (
+      <div className={styles.checkboxWithSelectWrapper}>
+        <CheckboxWithSelects
+          label={entry.label}
+          disabled={entry.disabled}
+          onChange={(selectedOptions: string[]) =>
+            onChangeHandler(entry.type, currentPath, selectedOptions)
+          }
+          selectedOptions={selectedOptions}
+          options={(entry.options as Option[]) || []}
+        />
+      </div>
+    );
+  };
+
+  const buildCheckboxWithRadios = (entry: AttributeWithOptions, path: Path) => {
+    const currentPath = [...path, entry.id];
+
+    const selectedOption: string = get(props.selectedData, currentPath, '');
+
+    return (
+      <div className={styles.checkboxWitRadiosWrapper}>
+        <CheckboxWithRadios
+          label={entry.label}
+          disabled={entry.disabled}
+          onChange={(selectedOption: string | number | boolean) =>
+            onChangeHandler(entry.type, currentPath, selectedOption)
+          }
+          selectedOption={selectedOption}
+          options={entry.options as RadioOptions}
+        />
+      </div>
+    );
+  };
+
+  const buildCheckboxGrid = (entry: AttributeWithOptions, path: Path) => {
+    const currentPath = [...path, entry.id];
+
+    const selectedOptions = get(props.selectedData, currentPath, []);
+
+    const gridOptions = [...entry.options];
+
+    const gridClone: CheckboxGridOption[] = [];
+
+    Object.values(gridOptions as CheckboxGridOption[]).map((option) => {
+      const optionClone = { ...option };
+      if (selectedOptions[option.id] !== undefined) {
+        optionClone.isChecked = selectedOptions[option.id];
+      } else if (optionClone.isChecked) {
+        onChangeHandler(entry.type, [...currentPath, optionClone.id], true);
+      }
+      gridClone.push(optionClone);
+    });
```

Comment:
Please do not use `.map` as `.forEach`! Also, please do not do side effects inside a `.map`!

If you want to call `onChangeHandler` for all checked values while building the checkboxGrid (will it result in calling a redux action for every isChecked?), then run a separate `forEach`:

Also, there was no point in declaring the `gridClone` variable as an empty array. `.map` function returns a new array.

```
gridOptions.forEach((option) => {
  if (option.isChecked) {
     onChangeHandler(entry.type, [...currentPath, option.id], true);
   }

gridOptions = gridOptions.map(option => selectedOptions[option.id]  ? {
  ...option,
   isChecked: selectedOptions[option.id]
} : option;
```

Reply:
>will it result in calling a redux action for every isChecked? 
It looks like it is supposed to dispatch many actions but it wasn't. Not sure why.
I've anyways updated this logic to call the onChange once all the options are processed.